## Boxoffice Mojo Web Scrape
An exercise in Absolute Frequency, Weighted Frequency and Relative Frequency by web scraping the all-time domestic box office results from box office mojo. This is modified from a Datacamp exercise.


Background on the metrics to be calculated:
- Absolute Frequency - number of times that a value appears, essentially, it is the number of times a particular thing happens. - in Movie Titles how many times does the word 'Star' appear

- Weighted Frequency - weights the individual value in terms of frequencies for each result in the 'value' content (overall frequency in the complete data set). - what is the value of 'Star' (how are we measuring)

- Relative frequency - the result of dividing the absolute frequency of a certain value by the total number of data. - Weighted Frequency divided by Absolute Frequency (times 'star' appears / value(measure) of star)


### Step One - gather the data


In [1]:
# import pandas, requests and Beautiful Soup

import pandas as pd
import requests
from bs4 import BeautifulSoup

In [2]:
# create the movies list including rank, title, studio lifetime gross and year.
# prints in increments of 10 for the web scrape progress. there are 150 pages to scrape
movie_list = []

for i in range(1, 156):
    if not i%10:
        print(i)
    page = 'http://www.boxofficemojo.com/alltime/domestic.htm?page=' + str(i) + '&p=.htm'
    resp = requests.get(page)
    soup = BeautifulSoup(resp.text, 'lxml')
    # trial and error to get the exact positions
    table_data = [x.text for x in soup.select('tr td')[11:511]]  
    # put every 5 values in a row
    temp_list = [table_data[i:i+5] for i in range(0, len(table_data[:-4]), 5)] 
    for temp in temp_list:
        movie_list.append(temp)

10
20
30
40
50
60
70
80
90
100
110
120
130
140
150


In [3]:
# review final list to ensure accuarcy
movie_list

[['1', 'Star Wars: The Force Awakens', 'BV', '$936,662,225', '2015'],
 ['2', 'Avatar', 'Fox', '$760,507,625', '2009^'],
 ['3', 'Black Panther', 'BV', '$700,059,566', '2018'],
 ['4', 'Avengers: Infinity War', 'BV', '$678,690,536', '2018'],
 ['5', 'Titanic', 'Par.', '$659,363,944', '1997^'],
 ['6', 'Jurassic World', 'Uni.', '$652,270,625', '2015'],
 ['7', "Marvel's The Avengers", 'BV', '$623,357,910', '2012'],
 ['8', 'Star Wars: The Last Jedi', 'BV', '$620,181,382', '2017'],
 ['9', 'Incredibles 2', 'BV', '$597,247,671', '2018'],
 ['10', 'The Dark Knight', 'WB', '$534,858,444', '2008^'],
 ['11', 'Rogue One: A Star Wars Story', 'BV', '$532,177,324', '2016'],
 ['12', 'Beauty and the Beast (2017)', 'BV', '$504,014,165', '2017'],
 ['13', 'Finding Dory', 'BV', '$486,295,561', '2016'],
 ['14',
  'Star Wars: Episode I - The Phantom Menace',
  'Fox',
  '$474,544,677',
  '1999^'],
 ['15', 'Star Wars', 'Fox', '$460,998,007', '1977^'],
 ['16', 'Avengers: Age of Ultron', 'BV', '$459,005,868', '2015']

In [4]:
# convert list to a Dataframe
df = pd.DataFrame.from_records(movie_list)
print(df.shape)
df.head(10)

(15500, 5)


0                             1     2             3      4
0   1  Star Wars: The Force Awakens    BV  $936,662,225   2015
1   2                        Avatar   Fox  $760,507,625  2009^
2   3                 Black Panther    BV  $700,059,566   2018
3   4        Avengers: Infinity War    BV  $678,690,536   2018
4   5                       Titanic  Par.  $659,363,944  1997^
5   6                Jurassic World  Uni.  $652,270,625   2015
6   7         Marvel's The Avengers    BV  $623,357,910   2012
7   8      Star Wars: The Last Jedi    BV  $620,181,382   2017
8   9                 Incredibles 2    BV  $597,247,671   2018
9  10               The Dark Knight    WB  $534,858,444  2008^

### Step Two - Review & Clean the data

In [5]:
# rename the columns
df.columns = ['rank', 'title', 'studio', 'lifetime_gross', 'year']

In [6]:
df.head()

rank                         title studio lifetime_gross   year
0    1  Star Wars: The Force Awakens     BV   $936,662,225   2015
1    2                        Avatar    Fox   $760,507,625  2009^
2    3                 Black Panther     BV   $700,059,566   2018
3    4        Avengers: Infinity War     BV   $678,690,536   2018
4    5                       Titanic   Par.   $659,363,944  1997^

In [7]:
# we have some characters to remove to 'clean' up: '$', ',' '^', ':'
# review the year column
df['year'].unique()

array(['2015', '2009^', '2018', '1997^', '2012', '2017', '2008^', '2016',
       '1999^', '1977^', '2004', '1982^', '2013', '2006', '1994^', '2010',
       '2002', '1993^', '2009', '2011', '2003^', '2005^', '2004^', '2014',
       '2002^', '2007', '2008', '2001', '2001^', '1983^', '1996^', '2003',
       '1999', '2005', '1980^', '1990', '2000', '1975', '1989', '1997',
       '1981^', '1984^', '1996', '1984', '1973^', '1993', '1991^', '1992',
       '1998', '1985', '1939^', '1995^', '2017^', '1978^', '1937^',
       '1995', '1986^', '1982', '1986', '1988', '1987', '1991', '1965',
       '1973', '2015^', '1994', '1961^', '1967^', '2010^', '1972^',
       '1978', '2013^', '1977', '1992^', '1974^', '1990^', '1981', '1976',
       '1974', '1989^', '1983', '1970', '1979', '1980', '1969', '1964^',
       '1942^', '1985^', '1955^', '1950^', '1953^', '1972', '1940^',
       '1979^', '1941^', '1988^', '1959', '1964', '1956', '1987^',
       '2000^', '1968^', '1963', '1998^', '1967', '1971', '195

In [8]:
# we have 'n/a' in the data, how many and where?
df[df['year'] == 'n/a']

rank                                        title studio  \
8088    8089  Warner Bros. 75th Anniversary Film Festival     WB   
8232    8233                Hum Aapke Dil Mein Rahte Hain   Eros   
8282    8283                       Purple Moon (Re-issue)  Mira.   
10598  10599                                     Amarcord   Jan.   

      lifetime_gross year  
8088        $741,855  n/a  
8232        $668,678  n/a  
8282        $640,945  n/a  
10598       $125,493  n/a

In [9]:
# not relevant so we will delete the 'n/a'
# find the index and remove the 'n/a' items that match
na_year_idx =  [i for i, x in enumerate(movie_list) if x[4] == 'n/a'] # get the indexes of the 'n/a' values
na_year_idx = set(na_year_idx)
final_list = [v for i, v in enumerate(movie_list) if i not in na_year_idx]

In [10]:
# double check to ensure we removed 4 items
len(final_list)

15496

In [11]:
# create dataframe
# remove special characters
# rename columns
# change data types where applicable

import re
regex = '|'.join(['\$', ',', '\^']) # list of special chars to remove

columns = ['rank', 'title', 'studio', 'lifetime_gross', 'year']

bom_df = pd.DataFrame({
    'rank': [int(x[0]) for x in final_list],
    'title': [x[1] for x in final_list],
    'studio': [x[2] for x in final_list],
    'lifetime_gross': [int(re.sub(regex, '', x[3])) for x in final_list],  
    'year': [int(re.sub(regex, '', str(x[4]))) for x in final_list],  }) 

print(bom_df.shape)
bom_df.head(10)

(15496, 5)


rank                         title studio  lifetime_gross  year
0     1  Star Wars: The Force Awakens     BV       936662225  2015
1     2                        Avatar    Fox       760507625  2009
2     3                 Black Panther     BV       700059566  2018
3     4        Avengers: Infinity War     BV       678690536  2018
4     5                       Titanic   Par.       659363944  1997
5     6                Jurassic World   Uni.       652270625  2015
6     7         Marvel's The Avengers     BV       623357910  2012
7     8      Star Wars: The Last Jedi     BV       620181382  2017
8     9                 Incredibles 2     BV       597247671  2018
9    10               The Dark Knight     WB       534858444  2008

In [12]:
# check datatypes
bom_df.dtypes

rank               int64
title             object
studio            object
lifetime_gross     int64
year               int64
dtype: object

In [13]:
# quick stats
bom_df.describe()

rank  lifetime_gross          year
count  15496.000000    1.549600e+04  15496.000000
mean    7750.228833    1.826879e+07   2003.388358
std     4475.123375    4.588610e+07     11.187391
min        1.000000    1.808000e+03   1921.000000
25%     3874.750000    6.218275e+04   1996.000000
50%     7748.500000    9.765930e+05   2006.000000
75%    11626.250000    1.544961e+07   2012.000000
max    15500.000000    9.366622e+08   2018.000000

In [14]:
# review the year
print(bom_df['year'].unique())
print()
print(bom_df['year'].max())
print(bom_df['year'].min())

[2015 2009 2018 1997 2012 2017 2008 2016 1999 1977 2004 1982 2013 2006
 1994 2010 2002 1993 2011 2003 2005 2014 2007 2001 1983 1996 1980 1990
 2000 1975 1989 1981 1984 1973 1991 1992 1998 1985 1939 1995 1978 1937
 1986 1988 1987 1965 1961 1967 1972 1974 1976 1970 1979 1969 1964 1942
 1955 1950 1953 1940 1941 1959 1956 1968 1963 1971 1962 1960 1954 1952
 1966 1957 1946 1945 1921 1958 1927 1926 1949 1947 1936 1931]

2018
1921


In [15]:
print(bom_df['lifetime_gross'].max())
print(bom_df['lifetime_gross'].min())

936662225
1808


15,496 movies 

Year ranges from 1921 - 2018

Life_time Gross ranges from \$1,788 to \$936,662,225

### Step Three - Calculate Absolute Frequency, Weighted Frequency and Relative Value of Words in Movie Titles

In [16]:
# to account for duplicates we will lowercase all words: example The and the are the same. 

from collections import defaultdict

def word_frequency(text_list, num_list, sep=None):
    word_freq = defaultdict(lambda: [0, 0])

    for text, num in zip(text_list, num_list):
        for word in text.split(sep=sep): 
            word_freq[word.lower()][0] += 1 
            word_freq[word.lower()][1] += num

    columns = {0: 'abs_freq', 1: 'wtd_freq'}

    abs_wtd_df = (pd.DataFrame.from_dict(word_freq, orient='index')
                 .rename(columns=columns )
                 .sort_values('abs_freq', ascending=False)
                 .assign(rel_value=lambda df: df['wtd_freq'] / df['abs_freq']).round())

    abs_wtd_df.insert(1, 'abs_perc', value=abs_wtd_df['abs_freq'] / abs_wtd_df['abs_freq'].sum())
    abs_wtd_df.insert(2, 'abs_perc_cum', abs_wtd_df['abs_perc'].cumsum())
    abs_wtd_df.insert(4, 'wtd_freq_perc', abs_wtd_df['wtd_freq'] / abs_wtd_df['wtd_freq'].sum())
    abs_wtd_df.insert(5, 'wtd_freq_perc_cum', abs_wtd_df['wtd_freq_perc'].cumsum())

    return abs_wtd_df
word_freq_df = word_frequency(bom_df['title'], bom_df['lifetime_gross'])
word_freq_df.head(10).style.bar(['abs_freq', 'wtd_freq', 'rel_value'], color='#edc993')

- So, the top words are stop words. 'the' is the top word and is in %10 of all movie titles.

- Let's refine and remove selected stopwords and recalulate:
'of','in', 'to', 'and', 'a', 'the', 'for', 'on', '&', 'is', 'at', 'it', 'from', 'with'


In [17]:
def word_frequency(text_list, num_list, sep=None, rm_words=('of','in', 'to', 'and', 'a', 'the', 
                                         'for', 'on', '&', 'is', 'at', 'it',
                                         'from', 'with')):  
    word_freq = defaultdict(lambda: [0, 0])

    for text, num in zip(text_list, num_list):
        for word in text.split(sep=sep): 
            # This should take care of ignoring the word if it's in the stop words
            if word.lower() in rm_words:  
                continue                  
            # .lower() makes sure we are not duplicating words
            word_freq[word.lower()][0] += 1  
            word_freq[word.lower()][1] += num

    columns = {0: 'abs_freq', 1: 'wtd_freq'}

    abs_wtd_df = (pd.DataFrame.from_dict(word_freq, orient='index')
                 .rename(columns=columns )
                 .sort_values('abs_freq', ascending=False)
                 .assign(rel_value=lambda df: df['wtd_freq'] / df['abs_freq']).round())

    abs_wtd_df.insert(1, 'abs_perc', value=abs_wtd_df['abs_freq'] / abs_wtd_df['abs_freq'].sum())
    abs_wtd_df.insert(2, 'abs_perc_cum', abs_wtd_df['abs_perc'].cumsum())
    abs_wtd_df.insert(4, 'wtd_freq_perc', abs_wtd_df['wtd_freq'] / abs_wtd_df['wtd_freq'].sum())
    abs_wtd_df.insert(5, 'wtd_freq_perc_cum', abs_wtd_df['wtd_freq_perc'].cumsum())

    abs_wtd_df = abs_wtd_df.reset_index().rename(columns={'index': 'word'})

    return abs_wtd_df
word_freq_df = word_frequency(bom_df['title'], bom_df['lifetime_gross'])
word_freq_df.head(10).style.bar(['abs_freq', 'wtd_freq', 'rel_value'], color='#edc993')

In [18]:
(word_freq_df.sort_values('wtd_freq', ascending=False)
 .head(10)
    .style.bar(['abs_freq', 'wtd_freq', 'rel_value'],
               color='#edc993'))

In [19]:
(word_freq_df.sort_values('rel_value', ascending=False)
 .head(10)
    .style.bar(['abs_freq', 'wtd_freq', 'rel_value'],
               color='#edc993'))

In [20]:
# abs_freq greater than 1
abs_greter_than_1 = word_freq_df[word_freq_df['abs_freq'] > 1]
(abs_greter_than_1.sort_values('rel_value', ascending=False)
 .head(10)
    .style.bar(['abs_freq', 'wtd_freq', 'rel_value'],
               color='#edc993'))

In [21]:
abs_greter_than_2 = word_freq_df[word_freq_df['abs_freq'] > 2]
(abs_greter_than_2.sort_values('rel_value', ascending=False)
 .head(10)
    .style.bar(['abs_freq', 'wtd_freq', 'rel_value'],
               color='#edc993'))

### Step Four - Analysis and Export


- Love is the most used word in movie titles. Out of 15,496 movies, Love appears in 210 movie titles followed by Man(197), My(191), 'I'(168) and '2'(159). 


- Love is not that high when it comes Weighted Frequency. In fact, Love is not even in the top ten. The number 2 is the top word (I know 2 is not a word but, in this analysis, '2' and '3' and 'i' and 'ii' are being used as words as they are part of movie titles). Weighted Frequency is calculating the sum of lifetime gross revenue for each movie the word appeared in the title. With this as the metric, 2's weighted frequency is worth \$10,062,046,737 - that's right 2 is in the title of movies whose sum is greater than 10 billion. This is almost double the next entry 'star' at \$5,588,247,468 - thanks Star Wars and Star Trek. 'Part' is interesting as it is a combination of 2 or 3 or ii which is the value of the 2nd and 3rd parts of a movie franchise. 


- With Relative Value, the words with an abs_freq of 1 are actually the lifetime gross of one movie, i.e. awakens is a word in the number one movie of all time 'Star Wars: The Force Awakens'. Ok but what about abs_freq is > 1? Similar findings with highly successful franchises or the second movie. Ok what about abs_freq greater than 2 - Again Similar findings but now 'wars' has the highest relative value.

In [22]:
# a way to sense check the words in titles and further explore other top words.
bom_df[bom_df['title'].str.contains('star | star', case=False)].head(10)

rank                                         title studio  \
0       1                  Star Wars: The Force Awakens     BV   
7       8                      Star Wars: The Last Jedi     BV   
10     11                  Rogue One: A Star Wars Story     BV   
13     14     Star Wars: Episode I - The Phantom Menace    Fox   
14     15                                     Star Wars    Fox   
36     37  Star Wars: Episode III - Revenge of the Sith    Fox   
69     70  Star Wars: Episode II - Attack of the Clones    Fox   
109   110                                     Star Trek   Par.   
144   145                       Star Trek Into Darkness   Par.   
171   172                       Solo: A Star Wars Story     BV   

     lifetime_gross  year  
0         936662225  2015  
7         620181382  2017  
10        532177324  2016  
13        474544677  1999  
14        460998007  1977  
36        380270577  2005  
69        310676740  2002  
109       257730019  2009  
144       228778661  2013  
171       213601143  2018

In [23]:
# what is the gross by studio
total_gross_studio = bom_df.groupby('studio')['lifetime_gross'].agg(sum).sort_values(ascending=False)
total_gross_studio.head()

studio
BV      41352429405
WB      37794700855
Uni.    33836240681
Fox     32855030953
Par.    28055922760
Name: lifetime_gross, dtype: int64

In [24]:
# number of movies by studio
total_movies = bom_df['studio'].value_counts()
total_movies.head()

WB      803
Uni.    691
Fox     617
Par.    591
BV      570
Name: studio, dtype: int64

In [25]:
# create a new dataframe, reset the index, rename the columns
new_df = pd.DataFrame([dict(total_gross_studio), dict(total_movies)]).transpose()
new_df = new_df.rename_axis('studio').reset_index()
new_df.columns = ['studio','lifetime_gross', 'number_movies']

In [26]:
# get the average total gross by studio
new_df.insert(3, 'avg_gross', value=new_df['lifetime_gross'] / new_df['number_movies'])

In [27]:
new_df.sort_values('avg_gross', ascending = False).head(10)

studio  lifetime_gross  number_movies     avg_gross
698      P/DW      4730766291             39  1.213017e+08
573       MFF        87178599              1  8.717860e+07
1014  WB (NL)      4799255502             59  8.134331e+07
142        BV     41352429405            570  7.254812e+07
260        DW      4283727041             60  7.139545e+07
271      Dis.      1432560292             21  6.821716e+07
915      Sum.      1663937182             31  5.367539e+07
32        AAP        53267000              1  5.326700e+07
364       Fox     32855030953            617  5.324964e+07
523      LG/S      2031545083             39  5.209090e+07

BV or Buena Vista is top in total gross \$’s but is 5th in the total number of movies. WB or Warner Brothers is 2nd in total gross \$’s but 1st in total number of movies. The Star Wars franchise really boosts BV. 

How about the average gross by studio?

1.  is Paramount (DreamWorks) (Transformers, Shrek, Madagascar etc)
2.  is MacGillivray Freeman Films a documentary on Everest IMAX(only) 
3.  is Warner Bros. (New Line) (The Hobbit, Sex And The City
4.  is BV or Disney (Starwars, Marvel movies, Toy Story)
5.  is Dreamworks (see above)

I could apply the same metrics with Absolute Frequency, Weighted Frequency and Relative frequency of the studios but I will save that for another day.

Exporting the files

In [28]:
word_freq_df.to_csv('word_freq_df.csv')
bom_df.to_csv('bom_df.csv')